stellargraph/mapper/mini_batch_node_generators.py

```diff
@@ -0,0 +1,361 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2018-2019 Data61, CSIRO
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Mappers to provide input data for the graph models in layers.
+
+"""
+__all__ = ["ClusterNodeGenerator", "ClusterNodeSequence"]
+
+import random
+import copy
+import numpy as np
+import networkx as nx
+from tensorflow.keras.utils import Sequence
+
+from scipy import sparse
+from ..core.graph import StellarGraphBase
+from ..core.utils import is_real_iterable
+
+
+class ClusterNodeGenerator:
+    """
+    A data generator for use with ClusterGCN models on homogeneous graphs, [1].
+
+    The supplied graph G should be a StellarGraph object that is ready for
+    machine learning. Currently the model requires node features to be available for all
+    nodes in the graph.
+    Use the :meth:`flow` method supplying the nodes and (optionally) targets
+    to get an object that can be used as a Keras data generator.
+
+    This generator will supply the features array and the adjacency matrix to a
+    mini-batch Keras graph ML model.
+
+    [1] `W. Chiang, X. Liu, S. Si, Y. Li, S. Bengio, C. Hsieh, 2019 <https://arxiv.org/abs/1905.07953>`_.
+
+    For more information, please see the ClusterGCN demo:
+        `<https://github.com/stellargraph/stellargraph/blob/master/demos/>`_
+
+    Args:
+        G (StellarGraphBase): a machine-learning StellarGraph-type graph
+        k (int): The number of clusters if parameter `clusters` is None. Otherwise it is ignored.
+        lam (float): The mixture coefficient for adjacency matrix normalisation.
+        clusters (list): a list of lists of node IDs such that each list corresponds to a cluster of nodes
+            in G. The clusters should be non-overlapping. If None, the G is clustered into k clusters.
+        name (str): an optional name of the generator
+    """
+
+    def __init__(self, G, k=1, q=1, lam=0.1, clusters=None, name=None):
+
+        if not isinstance(G, StellarGraphBase):
+            raise TypeError("Graph must be a StellarGraph object.")
+
+        self.graph = G
+        self.name = name
+        self.k = k
+        self.q = q  # The number of clusters to sample per mini-batch
+        self.lam = lam
+        self.clusters = clusters
+
+        if clusters:
+            self.k = len(clusters)
+
+        # Some error checking on the given parameter values
+        if lam < 0 or lam > 1 or not isinstance(lam, float):
```

Comment:
Type test should go before value tests.

Reply:
And technically, our newest convention is that failure of the type test should result in a `TypeError`.

Reply:
I'll update accordingly, but this newest convention is news to me. Where have we documented these conventions for testing?

Reply:
Not sure, but it was the end result of conversations between @adocherty and myself, which I gather also come from the Python convention regarding errors.